File: src/main/java/com/trustline/trustline/config/security/JWTConfig.java
package com.trustline.trustline.config.security;

import com.google.api.client.util.Value;
import com.trustline.trustline.appuser.model.User;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.util.Date;

@Service
public class JWTConfig {

    @Value("${jwt.signin.secret}")
    private String secret;


private SecretKey setKey(){
   return Keys.secretKeyFor(SignatureAlgorithm.HS256);
}

    public String generateToken(User user){
        return Jwts.builder()
                .setSubject(user.getEmail())
                .claim("roles", user.getRoles())
                .claim("status", user.getStatus())
                .claim("id", user.getId())
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + 60 * 60 *60 * 30))
                .signWith(setKey(), SignatureAlgorithm.HS256)
                .compact();
    }


    public String extractEmail(String token){
        return extractClaims(token).getSubject();
    }
    private Claims extractClaims(String token){
        return Jwts.parserBuilder()
                .setSigningKey(Keys.hmacShaKeyFor(secret.getBytes()))
                .build()
                .parseClaimsJws(token)
                .getBody();
    }

    public boolean isTokenValid(String token, UserDetails userDetails) {
        return extractEmail(token).equals(userDetails.getUsername()) && !isTokenExpired(token);
    }


    private Boolean isTokenExpired(String token) {
        return extractClaims(token).getExpiration().before(new Date(System.currentTimeMillis()));
    }

}

File: src/main/kotlin/trustline/config/security/JWTConfig.kt
package trustline.config.security;

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Service
import trustline.appuser.model.UserModel
import java.util.*
import javax.crypto.SecretKey

@Service
class JWTConfig(
    @Value("\${jwt.signin.secret}")
    private val secret: String
) {


    private fun setKey(): SecretKey =
        Keys.secretKeyFor(SignatureAlgorithm.HS256)

    fun generateToken(user: UserModel): String =
        Jwts.builder()
            .setSubject(user.email)
            .claim("roles", user.roles)
            .claim("status", user.status)
            .claim("id", user.id)
            .setIssuedAt(Date(System.currentTimeMillis()))
            .setExpiration(Date(System.currentTimeMillis() + 60 * 60 * 60 * 30))
            .signWith(setKey(), SignatureAlgorithm.HS256)
            .compact()

    fun extractEmail(token: String): String =
        extractClaims(token).subject

     fun extractClaims(token: String): Claims =
        Jwts.parserBuilder()
            .setSigningKey(Keys.hmacShaKeyFor(secret.toByteArray()))
            .build()
            .parseClaimsJws(token)
            .body

    fun isTokenValid(token: String, userDetails: UserDetails): Boolean =
        extractEmail(token) == userDetails.username && !isTokenExpired(token)

     fun isTokenExpired(token: String): Boolean =
        extractClaims(token).expiration.before(Date(System.currentTimeMillis()))
}
